src/sidebar/CustomModelBox.tsx

```diff
@@ -0,0 +1,84 @@
+import 'codemirror/lib/codemirror.css'
+import 'codemirror/addon/hint/show-hint.css'
+import 'codemirror/addon/lint/lint.css'
+// todonow: this belongs to this app and we should not take it from the demo...
+import 'custom-model-editor/demo/style.css'
+import styles from '@/sidebar/CustomModelBox.module.css'
+import React, { useEffect, useRef, useState } from 'react'
+import { create } from 'custom-model-editor/src/index'
+import Dispatcher from '@/stores/Dispatcher'
+import { SetCustomModel, SetCustomModelBoxEnabled } from '@/actions/Actions'
+import { CustomModel } from '@/stores/QueryStore'
+
+const initialCustomModel: CustomModel = {
+    speed: [],
+    priority: [],
+}
+
+export interface CustomModelBoxProps {
+    enabled: boolean
+    encodedValues: object[]
+}
+
+export default function CustomModelBox({ enabled, encodedValues }: CustomModelBoxProps) {
+    // todo: add types for custom model editor later
+    const [editor, setEditor] = useState<any>()
+    const divElement = useRef<HTMLDivElement | null>(null)
+    useEffect(() => {
+        // we start with empty categories. they will be set later using info
+        const instance = create({}, (element: Node) => divElement.current?.appendChild(element))
+        instance.value = JSON.stringify(initialCustomModel, null, 2)
+        setEditor(instance)
+        instance.validListener = (valid: boolean) => {
+            if (valid) Dispatcher.dispatch(new SetCustomModel(instance.jsonObj, true))
+            else {
+                try {
+                    const customModel = JSON.parse(instance.value)
+                    Dispatcher.dispatch(new SetCustomModel(customModel, false))
+                } catch (e) {
+                    Dispatcher.dispatch(new SetCustomModel(null, false))
+                }
+            }
+        }
+        Dispatcher.dispatch(new SetCustomModel(initialCustomModel, true))
```

Comment:
We need to update the app state here. This is the price we pay for not using the app's custom model state as the editor's backend. Afterwards we just update the `customModel` state whenever the editor evaluated the input's validity (see `validListener` above).